labgrid/driver/sshdriver.py

```diff
@@ -507,16 +507,35 @@ def _check_keepalive(self):
         return self._keepalive.poll() is None
 
     def _stop_keepalive(self):
+        def log_output(name, output):
+            if not output:
+                return
+
+            s = output.decode("utf-8", errors="ignore")
+            for line in s.splitlines():
+                self.logger.warning("Keepalive %s %s: %s", self.networkservice.address, name, line)
+
         assert self._keepalive is not None
 
         self.logger.debug('Stopping keepalive for %s', self.networkservice.address)
 
+        stdout = None
+        stderr = None
+
         try:
-            self._keepalive.communicate(timeout=60)
+            stdout, stderr = self._keepalive.communicate(timeout=60)
         except subprocess.TimeoutExpired:
             self._keepalive.kill()
-
-        try:
-            self._keepalive.wait(timeout=60)
+            try:
+                # Try again to get output
+                stdout, stderr = self._keepalive.communicate(timeout=60)
+            except subprocess.TimeoutExpired:
+                try:
+                    self._keepalive.wait(timeout=60)
+                except subprocess.TimeoutExpired:
+                    self.logger.warning("ssh keepalive for %s timed out during termination", self.networkservice.address)
         finally:
             self._keepalive = None
+
+        log_output("stdout", stdout)
+        log_output("stderr", stderr)
```

Comment:
As you said, normally no output is generated. Is it worth to distinguish stdout/stderr here? Maybe change the `subprocess.Popen` in `_start_keepalive()` (stdin isn't used):

```diff
         self._keepalive = subprocess.Popen(
             args,
-            stdin=subprocess.PIPE,
             stdout=subprocess.PIPE,
-            stderr=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
         )
 ```
If you also specify `encoding="utf-8"` there, you can probably also drop `log_output()` and do the logging here? The logging could even happen in the `try..except`, so you don't have to initialize the variables beforehand. This would slim down the patch even more.

Reply:
stdin has to be `subprocess.PIPE` because this is how the keep alive is terminated (closing stdin), but otherwise I think I've done all of these.